MantidQt/SliceViewer/src/PeakViewFactory.cpp

```diff
@@ -0,0 +1,219 @@
+{
+    std::vector<Mantid::Kernel::V3D> directions;
+    if (frame.name() == Mantid::Geometry::QSample::QSampleName) {
+        Mantid::Kernel::Matrix<double> goniometerMatrix
+            = peak.getGoniometerMatrix();
+        if (goniometerMatrix.Invert() != 0.0) {
+            directions
+                = ellipticalShape.getDirectionInSpecificFrame(goniometerMatrix);
+        } else {
+            directions = ellipticalShape.directions();
+        }
+    } else {
+
+        directions = ellipticalShape.directions();
+    }
+    return directions;
+}
+}
+
+namespace MantidQt
+{
+namespace SliceViewer
+{
+
+PeakViewFactory::PeakViewFactory(Mantid::API::IMDWorkspace_sptr mdWS,
+                                 Mantid::API::IPeaksWorkspace_sptr peaksWS,
+                                 QwtPlot *plot, QWidget *parent,
+                                 const int plotXIndex, const int plotYIndex,
+                                 const size_t colorNumber)
+    : PeakOverlayViewFactoryBase(plot, parent, plotXIndex, plotYIndex,
+                                 colorNumber),
+      m_mdWS(mdWS), m_peaksWS(peaksWS),
+      m_calculator(
+          std::
+              make_shared<Mantid::SliceViewer::EllipsoidPlaneSliceCalculator>())
+{
+    setForegroundAndBackgroundColors(colorNumber);
+}
+
+PeakViewFactory::~PeakViewFactory() {}
+
+boost::shared_ptr<PeakOverlayView> PeakViewFactory::createView(
+    PeaksPresenter *const presenter,
+    Mantid::Geometry::PeakTransform_const_sptr transform) const
+{
+  double largestEffectiveRadius = 0.0;
+  double currentEffectiveRadius = 0.0;
+    VecPeakRepresentation peakRepresentations(m_peaksWS->rowCount());
+    int index = 0;
+    for (auto &peakRepresentation : peakRepresentations) {
+        const Mantid::Geometry::IPeak &peak = m_peaksWS->getPeak(index);
+        auto position = transform->transformPeak(peak);
+        peakRepresentation
+            = createSinglePeakRepresentation(peak, position, transform);
+        // Get the largest radius of the data set
+        currentEffectiveRadius = peakRepresentation->getEffectiveRadius();
+        if (currentEffectiveRadius > largestEffectiveRadius) {
+          largestEffectiveRadius = currentEffectiveRadius;
+        }
+        ++index;
+    }
+
+    return boost::make_shared<PeakView>(
+        presenter, m_plot, m_parent, peakRepresentations, m_plotXIndex,
+        m_plotYIndex, m_foregroundColor, m_backgroundColor, largestEffectiveRadius);
+}
+
+PeakRepresentation_sptr PeakViewFactory::createSinglePeakRepresentation(
+    const Mantid::Geometry::IPeak &peak, Mantid::Kernel::V3D position,
+    Mantid::Geometry::PeakTransform_const_sptr transform) const
+{
+    // Available representations for this peaks: Cross, Sphere, Ellipsoid
+    const auto &peakShape = peak.getPeakShape();
+    const auto shapeName = peakShape.shapeName();
+
+    // Create the correct peak representation for the peak shape
+    PeakRepresentation_sptr peakRepresentation;
```

Comment:
Are there peak types which don't adhere to the `IPeak` interface? As long as it does and defines `shapeName` it should be fairly simple to extend the choice of peak representations.